docs/proposals/workflow.md

```diff
@@ -0,0 +1,255 @@
+    Finalizers map[string]string `json:"finalizers,omitempty"`
+    ExistenceDependencies []ObjectReference `json:"existenceDependencies,omitempty"`
+    ControllerRef *ObjectReference `json:"controllerRef,omitempty"`
+...
+}
+```
+
+### Recurring `Workflow` and `ScheduledJob`
+
+One of the major functionalities missing here is the ability to set a recurring `Workflow` (cron-like),
+similar to the `ScheduledJob` [#11980](https://github.com/kubernetes/kubernetes/pull/11980) for `Job`.
+If the scheduled job is able
+to support [various resources](https://github.com/kubernetes/kubernetes/pull/11980#discussion_r46729699)
+`Workflow` will benefit from the _schedule_ functionality of `Job`.
+
+
+### Graceful and immediate termination
+
+`Workflow` should support _graceful and immediate termination_ [#1535](https://github.com/kubernetes/kubernetes/issues/1535).
+
+
+## Implementation
+
+This proposal introduces a new REST resource `Workflow`. A `Workflow` is represented as a
+[graph](https://en.wikipedia.org/wiki/Graph_(mathematics)), more specifically as a DAG.
+Vertices of the graph represent steps of the workflow. The workflow steps are represented via a
+`WorkflowStep` resource.
+The edges of the graph are not represented explicitly - rather they are stored as a list of
+predecessors in each `WorkflowStep` (i.e. each node).
+
+
+### Workflow
+
+A new resource will be introduced in the API. A `Workflow` is a graph.
+In the simplest case it's a graph of `Job`s but it can also
+be a graph of other entities (for example cross-cluster objects or other `Workflow`s).
+
+
+```go
+// Workflow is a directed acyclic graph
+type Workflow struct {
+    unversioned.TypeMeta `json:",inline"`
+
+    // Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata.
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the expected behavior of a Workflow. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+    Spec WorkflowSpec `json:"spec,omitempty"`
+
+    // Status represents the current status of the Workflow. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+    Status WorkflowStatus `json:"status,omitempty"`
+}
+```
+
+
+#### `WorkflowSpec`
+
+```go
+// WorkflowSpec contains Workflow specification
+type WorkflowSpec struct {
+
+    // Optional duration in seconds the workflow needs to terminate gracefully. May be decreased in delete request.
+	// Value must be non-negative integer. The value zero indicates delete immediately.
+	// If this value is nil, the default grace period will be used instead.
+	// Set this value longer than the expected cleanup time for your workflow.
+    // If downstream resources (job, pod, etc.) define their TerminationGracePeriodSeconds
+    // the biggest is taken.
+	TerminationGracePeriodSeconds *int64 `json:"terminationGracePeriodSeconds,omitempty"`
```

Comment:
Sure: in a _workflowStep_ we can have basically two things:
1. a `JobTemplate` which will create a `Job` 
2. a reference to an already instantiated _Object_ (another _Workflow_ a remote ubernetes resource, etc)

For case 1 we need to handle _Graceful termination_, and adding this field will set the termination period for all subresources.